java_tools/tune-tools/src/test/java/com/rusefi/tune/TuneCanToolTest.java

```diff
@@ -0,0 +1,40 @@
+package com.rusefi.tune;
+
+import com.opensr5.ini.IniFileModel;
+import com.opensr5.ini.IniFileModelImpl;
+import com.rusefi.*;
+import com.rusefi.tools.tune.TuneCanTool;
+import com.rusefi.tune.xml.Msq;
+import com.rusefi.tune.xml.Page;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class TuneCanToolTest {
+    @Test
+    public void testGPPWMTuneParse() throws Exception {
+        Msq customOldTune = Msq.readTune(TuneReadWriteTest.TUNE_NAME);
+        Page dataPage = customOldTune.findPage();
+        assertFalse(dataPage.constant.isEmpty());
+
+        Msq lessOldDefaultTune = Msq.readTune(TuneCanToolTest.class.getResource("/simulator_tune-2023-06.xml").getFile());
+
+        IniFileModel ini = IniFileModelImpl.readIniFile(TuneReadWriteTest.TEST_INI);
+        assertEquals(256, ini.getBlockingFactor());
+        assertFalse(ini.getFieldsInUiOrder().isEmpty());
+
+        RootHolder.ROOT = "../../firmware/";
+
+        StringBuilder sb = TuneCanTool.getTunePatch(lessOldDefaultTune, customOldTune, ini, TuneReadWriteTest.TUNE_NAME, new StringBuilder(), null, "");
+
+        // tests GPPWM table bins
+        assertEquals("    // default \"Zero\"\n" +
+                     "    engineConfiguration->gppwm[0].loadAxis = GPPWM_Tps;\n" +
+                     "    // default \"Zero\"\n" +
+                     "    engineConfiguration->gppwm[1].loadAxis = GPPWM_Tps;\n" +
+                     "    // default \"Zero\"\n" +
+                     "    engineConfiguration->gppwm[2].loadAxis = GPPWM_Tps;\n" +
+                     "    // default \"Zero\"\n" +
+                     "    engineConfiguration->gppwm[3].loadAxis = GPPWM_Tps;\n", sb.substring(4565,4877));
```

Comment:
that's a very fragile unit test. in case of any changes of code generation it would be major pain to adjust those magic indices :(

please go with a less integrated test with smaller input and more focused output.